### PR TITLE
Refactor DatePicker component to react hooks and function component

### PR DESCRIPTION
**Pull request by amustaque97**
This PR is a part of issue #22890

## Description

It was already done as a part of different PR #22897. After code review
it looks like the original author is not able to find time to reply or
address review comments. I'm doing takeover from here and will take care
of PR until it gets merged.

Changes done as a part of PR
- Using react hook `useRef` in the component.
- Refactor from class component to function component.

## How has this been tested?

1. Edit page or post
2. Click on post schedule date
3. Change date
4. Publish post/page

## Types of changes

Code Quality - Refactor
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Review by ntsekouras — COMMENTED** [review]
Thanks for your contribution @amustaque97! 💯 

Code wise this looks good to me and I agree with [the comment you shared in core chat](https://wordpress.slack.com/archives/C02QB2JS7/p1637865119105900):

>After refactoring class component to function component few of the test cases have started failing. I have raised a PR. After looking at the test cases I believe we should remove those test cases because we should test component behaviour, not implementation details. What do you think?

I think @ciampo could share his thoughts about the way forward for the test changes, as he have worked a ton in components and we try to be as consistent as possible.

**Comment by ciampo** [comment]
Thank you @ntsekouras for the ping!

Hey @amustaque97, thank you for picking up this piece of work! I'm currently working on some `DatePicker` related changes in #36878, so maybe it's better to wait until that PR is merged, and rebase this PR to include those changes.

In the meantime, I'll give this PR a better look next week!

**Comment by ciampo** [comment]
Alright —  #36878 has been merged!

@amustaque97 , would you mind rebasing this PR on top of `trunk` in order to include the latest changes to the component?

I also 100% agree with this comment:

> After looking at the test cases I believe we should remove those test cases because we should test component behaviour, not implementation details. What do you think?

~Although I'm curious to know why tests are failing, since I didn't expect code changes in this PR to cause them to fail.~ Nevermind, I just realised that they're expecting certain functions to exist on the component class's instance!

Code changes so far look good, but I'll go through a detailed code review after the rebase if that's ok! Thank you for working on this once again!

**Comment by amustaque97** [comment]
> I just realised that they're expecting certain functions to exist on the component class's instance!

Yeah right, that's why, we should focus on overall component behavior rather than its internal functions.

I have taken merge from trunk branch. Please take a look once GitHub actions is complete.

File: packages/components/src/date-time/date.js
Comment (ciampo) [inline packages/components/src/date-time/date.js:97]: Shouldn't `keepFocusInside` be defined _before_ `onMonthPreviewedHandler`?

File: packages/components/src/date-time/date.js
Comment (ciampo) [inline packages/components/src/date-time/date.js:83]: A bit of a nit (and one may argue out of scope), but I noticed that the `onMonthPreviewed` prop isn't listed in the [README](https://github.com/WordPress/gutenberg/blob/cb8c98adc9a9071b2f6151499b6cc53d4900df56/packages/components/src/date-time/README.md#props) — would you mind adding it?

**Review by ciampo — COMMENTED** [review]
Overall, code changes look good — I've only added a couple of inline comments.

Regarding the failing unit tests, we definitely have to rewrite them, if anything to make them pass again. **I personally see 3 alternatives here**, each one representing a better solution than the previous one (but also a larger effort):

1. Rewrite the tests with the minimal changes to make them pass (i.e. not relying on the component being a Class component)
2. Rewrite part of the tests in order to test the component behaviour (instead of implementation details)
3. Do the same suggested in option 2, but also switch from Enzyme to React Testing Library

I appreciate that you may not have a lot of time to spend on this PR and/or you'd like to get it merged ASAP, and therefore **I'd be ok with any of the 3 options above**.

Finally, could you also add an entry to the [**CHANGELOG**](https://github.com/WordPress/gutenberg/blob/46f6a0dfe0aa784f2eb20ea120b5e661b66d76e8/packages/components/CHANGELOG.md) under the unreleased "Enhancements" section ?

Thank you!

File: packages/components/src/date-time/date.js
Comment (ntsekouras) [reply packages/components/src/date-time/date.js:97]: `onMonthPreviewedHandler` isn't called until all the code has run, so there isn't any TDZ(temporal dead zone) issue here.

File: packages/components/src/date-time/date.js
Comment (amustaque97) [reply packages/components/src/date-time/date.js:83]: I have added `onMonthPreviewed` props in the README.md file. Please check it once, I know I may have not written enough info but I tried my best to update the docs and not to confuse users. 😅

**Comment by amustaque97** [comment]
Removing `getMomentDate` test cases. Since in real word scenario, in UI we never deal with `getMomentDate` directly so in the test case we cannot write some event or simulate any behaviour that will give access to `getMomentDate` method. Now if we think of taking help of `currentDate` props then test cases are already in the place.

If you have any better idea please share. I would love to implement them.

**Comment by mirka** [comment]
Great refactor!

> Removing `getMomentDate` test cases. [...] If you have any better idea please share. I would love to implement them.

Since those tests are more like a unit test of the pure function `getMomentDate` to begin with (and doesn't even seem to be intended to test any component behavior at all), I could see the argument to move `getMomentDate` out of the `DatePicker` function and export it separately as a named export to test. But the tests aren't particularly complex and the behavior seems pretty much covered by the actual component-based `currentDate` tests like you say. So personally I'm fine with deletion too if you're happy with it!

Are we ready for final review?

**Comment by amustaque97** [comment]
Yeah, sure we can have a final round of review. :smile:

File: packages/components/CHANGELOG.md
Comment (ciampo) [inline packages/components/CHANGELOG.md]: Aligning white space to the other entries

```suggestion
-   Refactor `DateTime` class component to functional component ([#36835](https://github.com/WordPress/gutenberg/pull/36835))
```

File: packages/components/src/date-time/README.md
Comment (ciampo) [inline packages/components/src/date-time/README.md]:
```suggestion
A callback invoked when selecting the previous/next month in the date picker. The callback receives the new month date in the ISO format as an argument.
```

**Review by ciampo — COMMENTED** [review]
>  I could see the argument to move getMomentDate out of the DatePicker function and export it separately as a named export to test.\

This would be my favourite approach, especially since we already have those unit tests.

I went ahead on my local branch and made these changes, which seem to work well:
- moved `getMomentDate` to a separate `utils.js` file
- moved the `getMomentDate` tests under a new file, eg `test/utils.js` file
- changed the `getMomentDate` unit tests to avoid rendering `DatePicker` and using enzyme

<details>
<summary>Click to show diff</summary>

```diff
diff --git a/packages/components/src/date-time/date.js b/packages/components/src/date-time/date.js
index 9121fe0264..fd6869ac84 100644
--- a/packages/components/src/date-time/date.js
+++ b/packages/components/src/date-time/date.js
@@ -14,6 +14,11 @@ import DayPickerSingleDateController from 'react-dates/lib/components/DayPickerS
 import { useEffect, useRef } from '@wordpress/element';
 import { isRTL, _n, sprintf } from '@wordpress/i18n';
 
+/**
+ * Internal dependencies
+ */
+import { getMomentDate } from './utils';
+
 /**
  * Module Constants
  */
@@ -129,20 +134,6 @@ function DatePicker( {
 		keepFocusInside();
 	};
 
-	/**
-	 * Create a Moment object from a date string. With no date supplied, default to a Moment
-	 * object representing now. If a null value is passed, return a null value.
-	 *
-	 * @param {?string} date Date representing the currently selected date or null to signify no selection.
-	 * @return {?moment.Moment} Moment object for selected date or null.
-	 */
-	const getMomentDate = ( date ) => {
-		if ( null === date ) {
-			return null;
-		}
-		return date ? moment( date ) : moment();
-	};
-
 	const getEventsPerDay = ( day ) => {
 		if ( ! events?.length ) {
 			return [];
diff --git a/packages/components/src/date-time/test/utils.js b/packages/components/src/date-time/test/utils.js
new file mode 100644
index 0000000000..0f7d7a78ae
--- /dev/null
+++ b/packages/components/src/date-time/test/utils.js
@@ -0,0 +1,32 @@
+/**
+ * External dependencies
+ */
+import moment from 'moment';
+
+/**
+ * Internal dependencies
+ */
+import { getMomentDate } from '../utils';
+
+describe( 'getMomentDate', () => {
+	it( 'should return a Moment object representing a given date string', () => {
+		const currentDate = '1986-10-18T23:00:00';
+		const momentDate = getMomentDate( currentDate );
+
+		expect( moment.isMoment( momentDate ) ).toBe( true );
+		expect( momentDate.isSame( moment( currentDate ) ) ).toBe( true );
+	} );
+
+	it( 'should return null when given a null argument', () => {
+		const currentDate = null;
+		const momentDate = getMomentDate( currentDate );
+
+		expect( momentDate ).toBeNull();
+	} );
+
+	it( 'should return a Moment object representing now when given an undefined argument', () => {
+		const momentDate = getMomentDate();
+
+		expect( moment.isMoment( momentDate ) ).toBe( true );
+	} );
+} );
diff --git a/packages/components/src/date-time/utils.js b/packages/components/src/date-time/utils.js
new file mode 100644
index 0000000000..857b831ebb
--- /dev/null
+++ b/packages/components/src/date-time/utils.js
@@ -0,0 +1,18 @@
+/**
+ * External dependencies
+ */
+import moment from 'moment';
+
+/**
+ * Create a Moment object from a date string. With no date supplied, default to a Moment
+ * object representing now. If a null value is passed, return a null value.
+ *
+ * @param {?string} date Date representing the currently selected date or null to signify no selection.
+ * @return {?moment.Moment} Moment object for selected date or null.
+ */
+export const getMomentDate = ( date ) => {
+	if ( null === date ) {
+		return null;
+	}
+	return date ? moment( date ) : moment();
+};

```
</details>

@amustaque97 , do you think you'd be able to apply the same changes to this branch?

---

The actual refactor from class to functional component looks good to me.

I also noticed that you updated the tests with the minimal set of changes necessary to make them pass (i.e the option no. 1 from my [previous comment](https://github.com/WordPress/gutenberg/pull/36835#pullrequestreview-819304035)) — I take it that you don't intend to make further changes to the tests (as suggested by options 2 and 3) ? I'm asking just to confirm.

File: packages/components/CHANGELOG.md
Comment (amustaque97) [reply packages/components/CHANGELOG.md]: OOPs silly mistakes. 😛

**Comment by amustaque97** [comment]
> I take it that you don't intend to make further changes to the tests (as suggested by options 2 and 3) ? I'm asking just to confirm

Yes, I don't plan as of now 😅 . I'm not getting enough time to up-skill myself to learn about react-testing-library and make those changes. My suggestion is that we can create a separate issue and assign it to me. Maybe in coming new year holidays I can plan to submit a new PR `migrating existing test cases to react-testing-library`.

Let me know if that make sense to you  😄

**Review by ciampo — APPROVED** [review]
Thank you @amustaque97 for your work, these changes LGTM 🚀 

> Yes, I don't plan as of now 😅 . I'm not getting enough time to up-skill myself to learn about react-testing-library and make those changes. My suggestion is that we can create a separate issue and assign it to me. Maybe in coming new year holidays I can plan to submit a new PR migrating existing test cases to react-testing-library.
> 
> Let me know if that make sense to you 😄

That sounds like a great plan! Please do ping me and @mirka if/when you decide to submit a PR about that — we're always happy to help (also in case you are not sure about certain React Testing Library patterns!)

**Comment by ntsekouras** [comment]
Great work here @amustaque97 ! Also many thanks to the reviewers! 💯

**Pull request merged**
